File: Programmers/etc/주식가격.py
def solution(prices):
    l = len(prices)
    answer = [l-i-1 for i in range(len(prices))]
    cand_keys = []
    cand_idx = {}

    def upper_bound(s, e, p):
        if s >= e:
            return s
        mid = (s + e) // 2
        if cand_keys[mid] <= p:
            return upper_bound(mid + 1, e, p)
        else:
            return upper_bound(s, mid, p)

    for idx, price in enumerate(prices):
        if cand_keys and cand_keys[-1] > price:
            # 이분 탐색
            target = upper_bound(0,len(cand_keys)-1,price)
            for i in range(len(cand_keys)-1,target-1,-1):
                key = cand_keys[i]
                for j in cand_idx[key]:
                    diff = idx - j
                    answer[j] = diff
                del cand_idx[key]
                cand_keys.pop()
        if price in cand_idx:
            cand_idx[price].append(idx)
        else:
            cand_keys.append(price)
            cand_idx[price] = [idx]

    return answer

# prices	return
ex1 = [1, 2, 3, 2, 3] # [4, 3, 1, 1, 0]
ex2 = [5, 4, 3, 2, 1]
print(solution(ex2)) # [1, 1, 1, 1, 0]

File: Programmers/js/주식가격.js
function solution(prices) {
    const pricesSize = prices.length;
    const answer = prices.map((_, idx) => pricesSize - idx - 1);
    const remain = [];
    const remainIdx = [];

    const upperBound = (left, right, key) => {
        if (left >= right) {
            return right;
        }
        const middle = ~~((left + right) / 2);
        if (remain[middle] <= key) {
            return upperBound(middle + 1, right, key);
        } else {
            return upperBound(left, middle);
        }
    };

    prices.forEach((price, curr) => {
        const remainSize = remain.length;
        if (remainSize > 0 && remain[remainSize - 1] > price) {
            const targetIdx = upperBound(0, remainSize - 1, price);
            for (let prev = remainSize - 1; prev >= targetIdx; prev--) {
                answer[remainIdx[prev]] = curr - remainIdx[prev];
                remain.pop();
                remainIdx.pop();
            }
        }
        remain.push(price);
        remainIdx.push(curr);
    });

    return answer;
}

console.log(solution([1, 2, 3, 2, 3]));
console.log(solution([5, 4, 3, 2, 1]));
console.log(solution([3, 5, 6, 6, 5, 1]));